File: leetcode/problems/0001-two-sum/python/v1-solution.py
"""
LeetCode 1. Two Sum [Easy]
Link: https://leetcode.cn/problems/two-sum (source: leetcode.cn)
Tags: Array, Hash Table

Problem:
给定一个整数数组 `nums`&nbsp;和一个整数目标值 `target`，请你在该数组中找出 和为目标值 `target`&nbsp; 的那&nbsp;两个&nbsp;整数，并返回它们的数组下标。

你可以假设每种输入只会对应一个答案，并且你不能使用两次相同的元素。

你可以按任意顺序返回答案。

&nbsp;

示例 1：

输入：nums = [2,7,11,15], target = 9
输出：[0,1]
解释：因为 nums[0] + nums[1] == 9 ，返回 [0, 1] 。

示例 2：

输入：nums = [3,2,4], target = 6
输出：[1,2]

示例 3：

输入：nums = [3,3], target = 6
输出：[0,1]

&nbsp;

提示：

	- `2 &lt;= nums.length &lt;= 104`
	- `-109 &lt;= nums[i] &lt;= 109`
	- `-109 &lt;= target &lt;= 109`
	- 只会存在一个有效答案

&nbsp;

进阶：你可以想出一个时间复杂度小于 `O(n2)` 的算法吗？

Approach: TODO
Time: O(?), Space: O(?)
"""
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        

File: leetcode/problems/0283-move-zeroes/cpp/v1-solution.cpp
/*
 * LeetCode 283. Move Zeroes (Easy)
 * Link: https://leetcode.cn/problems/move-zeroes
 * Tags: Array, Two Pointers
 * Approach: TODO | Time: O(?) | Space: O(?)
 */

 /*
 给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。
请注意 ，必须在不复制数组的情况下原地对数组进行操作。
 */
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // TODO: 实现你的解法
    void moveZeroes(vector<int>& nums) {
        int n = nums.size();
        int j = 0; // 指向下一个非零元素应该放置的位置
        for (int i = 0; i < n; ++i) {
            if (nums[i] != 0) {
                if (i != j) {
                    swap(nums[i], nums[j]);
                }
                ++j;
            }
        }
    }
};

// ===== 本地简单测试 =====
// g++ -std=c++17 -O2 -pipe -static -s -o main v1-solution.cpp && ./main
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    // TODO: 添加你的本地测试
    Solution sol;
    vector<int> nums = {0, 1, 0, 3, 12};
    sol.moveZeroes(nums);
    for (int num : nums) {
        cout << num << " ";
    }
    return 0;
}
